### ISSUE TITLE
Options make the state of the object unclear

### ISSUE THREAD

**Issue by jodastephen** [maggiepint/proposal-temporal#6]
While I understand the desire for variation, I'm not conviced that options are the right approach.  A key part of having speific types for parts of the date/time problem space is that the types are *values*. Specifically, this means that given two `LocalDateTime` instances, if they are equal then they are also substitutable.

Consider this:

    var options1 = {
        resolver: (mapping) => ({
            skipped: mapping.forwardShifted(),
            ambiguous: mapping.firstOccurrence()
        })
    };
    var options2 = {
        resolver: (mapping) => ({
            skipped: mapping.nextValid(),
            ambiguous: mapping.firstOccurrence()
        })
    };
    var zdt1 = new ZonedDateTime('America/Los_Angeles', 2017, 3, 10, 8, 0, options1);
    var zdt2 = new ZonedDateTime('America/Los_Angeles', 2017, 3, 10, 8, 0, options2);
    // but is zdt1 equal to zdt2?

My assumption is that `zdt1` is not equal to `zdt2`. Yet this is plainly silly - both represent the same point in time, with the same time-zone. The only thing that differs is that they have been setup to react differently in certain circumstances, but that information is not really part of the state of the date-time, instead it is part of the business logic of the application. As such, the resolver must not be part of the state of `ZonedDateTime`, but should be passed in as a parameter when needed.

As further evidence of this, the date-time objects should have be able to be compared with a natural order, and in Java that order should be compatible with equals(). See the definition of equals() and compareTo() on `ZonedDateTime` in Java 8 to undertand how this works. It is also worth considering serialization, ie. how you can't really send a `ZonedDateTime` across a network as a string very easily if it contains additional information like a resolver.

In summary, the state of a date-time object has to be carefully controlled to just be the date, time, offset and time-zone. Resolvers don't make sense to be part of the state.

**Comment by mj1856** [comment]
Yep, I can't argue with your logic.  Options should be moved to functions that need them, so they don't need to be part of the state of the object.

**Comment by maggiepint** [comment]
The logic is quite correct, but I actually don't know how well this will play with JS. Keep in mind, we don't have overloads by type in the same way that C# and Java do. That limits one's options for how things can be overloaded.

**Comment by maggiepint** [comment]
This is why options objects are so popular in JavaScript in the first place. Now, we could move the options objects to the methods for computations, but no matter what, you're dealing in this space.

I think if you go down this road, you're looking at factory methods instead of the new keyword for a lot of stuff.

**Comment by mj1856** [comment]
Yeah, I think the right thing is that when we have options, they apply to the operation of a function, but do not become part of the state of the object.  We could indeed use them in the constructor still, for example:

```js
var ldt = new temporal.LocalDateTime(1438, 7, 15, 0, 0, { calendar: 'islamic'});
```

This wouldn't set the object to be islamic, but rather just interpret the input parameters as islamic and then discard the calendar information.  One wonders if this is any more or less clear as a factory function than as a constructor.

**Comment by jodastephen** [comment]
As I suggested in a different thread, I think it is clearer to start with the calendar system when working with non-ISO dates - ie. the calendar system object (chronology) is the primary object. Along the lines of:

    LocalDate date = IslamicChronology.INSTANCE.date(1438, 7, 15);
    int year = date.year(IslamicChronology.INSTANCE);
    int month = date.month(IslamicChronology.INSTANCE);
    int day = date.dayOfMonth(IslamicChronology.INSTANCE);

This works well because the abstraction is orthogonal to the concepts of date/time. A user can pass around the calendar system they want to use in the UI without affecting the business logic date/time objects.

**Comment by maggiepint** [comment]
Options are dead on constructors. Agree they're weird in that context.
